fixed viewport constraint behavior when viewer is rotated

pearcetm opened the pull request:
Related to https://github.com/openseadragon/openseadragon/pull/2041, fixes https://github.com/openseadragon/openseadragon/issues/2040

As far as I can tell, this works as expected with different values for `visibilityRatio`, viewport rotations, etc. I don't see a reason it should break anything, but... it would be good to have others test it out, of course.

One thing I haven't changed, but I'm wondering about, is why the `_raiseConstraintsEvent()` is called every time `applyConstraints()` is called, even if no constraints were actually applied... is this desirable? It basically means every navigation event raises the `constrain` event, even if no boundary was hit.

iangilman reviewed (commented):
This looks like a nice clean solution! 

The one odd thing is adding the extra properties to the `getConstrainedBounds`, since it's now technically more than a Rect. That said, it's cleaner than changing the return value more dramatically, and I don't suppose people will have problems with those extra properties. Still, might be worth updating the documentation of `getConstrainedBounds` to make it explicit. I mean, I suppose we could leave them as "private" but it seems like they might be useful additions to the API. 

Regarding the constraints event, that's a good point... I think it would be an improvement if it only happened when it hit the constraints. That would be a change in behavior, but I doubt any existing app is relying on getting that event in all cases. 

@AaronPlave I'd love to get your thoughts if you can take a look at this!

pearcetm commented:
@iangilman Good point about the documentation, I'd not really considered that this was part of a public API. I've added documentation of the new members. As you note, adding these properties is kind of strange, but I felt the same as you, that it was cleaner than a big revamp of the returned value - and this way, it's not a breaking change like it would be to return a structure with the bounds and flags as separate properties.

I would also note that I modified the old way of checking whether constraints had been applied (comparing the `x` and `y` properties of the original vs constrained bounds) because in my testing, that was breaking due to floating point math any time there was a rotation, even if the constraints aren't actually hit. This seemed cleaner and safer.

One thing I haven't tested out yet, because I really don't know what the intended use-case is, is the block of code at line 685 in `_fitBounds` when the `constraints` flag in the options argument is `true`. I don't think this should break anything that was previously working (since I don't think it was ever really working for non-zero rotations), but it would be good to test out and fix any bugs if needed...

pearcetm commented:
@iangilman I updated the relevant unit tests here in this PR. As far as I can tell, they are now working correctly. There was another test - I think `TileSource.animate` - that seemed to be failing because of floating-point issues. The test does not have epsilon values included for the asserts; I tweaked the implementation of `Viewport._applyBoundaryConstraints` to return the unmodified bounds if no constraints were hit, which should avoid small errors introduced by rotating and un-rotating the bounds. Perhaps eventually it'd be worth updating this unit test to allow floating point errors to pass the test though.

AaronPlave commented:
Excited that you took this on! The solution is working nicely in my testing, looking into one small quirk with zooming into the corner areas when rotated 90 degrees where the zoom goes in the wrong direction but I'll try again with an isolated test outside of my actual application, it's quite possible that this issue only affects my setup.

iangilman commented:
@pearcetm: 

> One thing I haven't tested out yet, because I really don't know what the intended use-case is, is the block of code at line 685 in _fitBounds when the constraints flag in the options argument is true. I don't think this should break anything that was previously working (since I don't think it was ever really working for non-zero rotations), but it would be good to test out and fix any bugs if needed...

Cool... That's called from `fitBoundsWithConstraints` which is there in case you want to enforce the constraints when you have it zoom/pan into a specific location. Should be pretty easy to set up a test for that... Let me know if you want pointers. 

@AaronPlave:

> Excited that you took this on! The solution is working nicely in my testing, looking into one small quirk with zooming into the corner areas when rotated 90 degrees where the zoom goes in the wrong direction but I'll try again with an isolated test outside of my actual application, it's quite possible that this issue only affects my setup.

Wonderful! Thank you for testing. It would be great to hear if that issue you found is something we need to address.

pearcetm commented on src/viewport.js:
```diff
@@ -514,34 +514,37 @@ $.Viewport.prototype = {
-                bounds.width,
-                bounds.height);
+    _applyBoundaryConstraints: function(bounds) {
+        var newBounds = this.viewportToViewerElementRectangle(bounds).getBoundingBox();
```
@iangilman Yes, exactly. The viewport rectangle needs to be rotated to match the viewer element, because the logic of the constraints is (and always was) implemented in terms of `x` and `y` directly. The rotated viewport bounds does not have the same `x` value for `topLeft` and `bottomLeft`, for example, because the corners are in the coordinate frame of the rotated image. If the logic was implemented as vector math relative to the line segments that define the edges of the rotated viewport rectangle, that would also work, but would be a huge refactor. In the previous implementation, which did not convert to viewer element coordinates, this was the reason for the unintuitive angled bouncing for non-90-degree rotations, and for the improper dimensions of the bounding area for 90-degree rotations.

Getting the bounding box for the rotated content is pretty obvious - we want to compute constraints from the min/max coordinates of the element versus the edges of the visible area. The reason for getting the bounding box of the viewport itself is less intuitive (isn't it already a rectangle?). The reason is that for some rotations, the coordinates of the rotated rectangle were automatically adjusting and making `topLeft` etc. incorrect, and the bounding box always is correct.

iangilman replied:
Awesome... Makes good sense! Thank you for the thorough explanation. This is of course all part of the historic record now :)

pearcetm commented on src/viewport.js:
```diff
@@ -794,7 +795,9 @@ $.Viewport.prototype = {
-     * @returns {OpenSeadragon.Viewport} Chainable.
+     * @returns {OpenSeadragon.Rect} The bounds after applying constraints. The returned $.Rect contains additonal
+     *                               properties xConstrained, yConstrained, constraintsApplied indicating the status
+     *                               of the constraining operation; x and y are in the viewer element coordinate system.
```
@iangilman See my response to your other comment - the constraints are actually in viewer element coordinates, not viewport coordinates. Conceptually, viewport coordinates are bound to the image; the constraints are bound to the viewer area. The "viewport" terminology has always seemed a bit counterintuitive to me...

pearcetm replied:
@iangilman Reading your comment again, I think I should clarify further: the bounds themselves _are_ in viewport coordinates. The viewer element coordinates, however, relate to the `xConstrained` and `yConstrained` flags. Specifically, `xConstrained` means the left or right side of the viewer element stopped the image from moving, and `yConstrained` means the top or bottom stopped it from moving - regardless of the current angle of rotation of the viewport.

pearcetm replied:
Updated the comment description, hopefully to be more clear

pearcetm commented:
@iangilman Thanks for pointing me in the right direction to look into `fitBoundsWithConstraints`. I found an old demo which appears that it probably hadn't worked in a long time, but it was a good starting point. Once I updated the demo (which can be loaded from http://localhost:8000/test/demo/fitboundswithconstraints.html when running the development environment), it was apparent that there were definitely problems with that functionality.

I think the changes I've now included have everything working right. I'd appreciate if you could take a look (and @AaronPlave too if possible).

pearcetm commented:
@iangilman Good point; the list was a carryover from the original demo page. I've updated the demo, and the comment for `getConstrainedBounds`

iangilman commented on src/viewport.js:
```diff
@@ -801,10 +801,10 @@ $.Viewport.prototype = {
-     *                               whether the viewport bounds were modified by the constraints of the viewer rectangle,
-     *                               and in which dimension(s).
+     * @returns {OpenSeadragon.Rect} The bounds in viewport coordinates after applying constraints. The returned $.Rect
+     *                               contains additonal properties constraintsApplied, xConstrained and yConstrained.
```
Typo: `additonal` should be `additional`. Sorry I didn't catch this before!

pearcetm replied:
@iangilman fixed.

iangilman reviewed (approved):
Looks great! Thank you for fixing up the demo! 

I think this is ready to go (with one super tiny nitpick), but I'd like to give @AaronPlave a chance to look into the issue he found.

AaronPlave commented:
Sorry about the delay, managed to reproduce this relatively minor issue on the demo page of this branch. Not sure if this issue is something introduced by this PR since one couldn't test this exact functionality on the master branch due without the fixes in this PR. I can't seem to upload any file to github PR comments right now so unfortunately need to use a temporary video link https://tempclip.com/en/E8NihPy9oPpu4XW/watch for the screencap of the issue. But to repro locally you can go to the fitBoundsWithConstraints demo, zoom in to the top left corner, rotate the image by 90º, and then zoom in to the top left corner (might have to hide the control buttons). The zoom in the top left corner on the rotated image does not zoom into the cursor and goes off in a different direction.

iangilman commented:
@AaronPlave Sure enough! Thank you for catching this! I can reproduce it, and I've tried it without the `visibilityRatio: 1.0` to validate that it is the constraints that are causing the problem. 

@pearcetm Can you take a look?

pearcetm commented:
Thanks for finding this @AaronPlave.

This is weird. I've done some digging, but I haven't sorted out yet what's going on.

It seems that when `degrees==0`, `viewportToViewerElementRectangle(bounds)` returns a `$.Rect` with `x==0, y==0, width==800, height==500` during all boundary checks, as I think it should. For non-zero rotations, though, the bounds rectangle when converted to viewer element `x, y` coordinates are NOT `0, 0` during animations. After a pan animation finishes, it returns to (close to) `0, 0`.

I'm not sure what to make of this yet. It doesn't seem right, though, and may indicate a deeper issue with how coordinate frames/transformations are implemented. @iangilman any thoughts? I'll keep digging in the meanwhile.

pearcetm commented:
I think I figured it out - it seems that rotation was not being taken into account within `Viewport.getCenter()` which was causing all sorts of weirdness. It seems to be working now, but please test! Hopefully fixing this doesn't break anything else that relies on `getCenter()`...

pearcetm commented:
@iangilman I am wondering if there are other places that rotated viewports aren't handled correctly (or handled at all), especially during spring-based animations. During animating rotation itself, for example... I'll take a look but thought you might have some good ideas of potential problem areas.

iangilman commented:
@pearcetm Looking good to me! I'm not surprised there are places like that that aren't taking rotation into account. I'm not sure what to look for in terms of negative effects. I did a search for `getCenter` in the code, and of course it's used in a lot of places (both directly and indirectly). The fact that the change hasn't broken any tests seems like a good sign. 

@AaronPlave It looks like the bug is fixed; do you want to take another look?

iangilman commented:
@AaronPlave Not to be a pest, but it would be great if you could take a look at the latest to verify that it fixes the issue you found! I'm hoping to merge this patch this week so we can do a new release :)

AaronPlave commented:
I'll take a look now! I need to sort out my github notifications, this keeps getting lost in the noise.

AaronPlave commented:
@iangilman @pearcetm the fix looks good to me, nicely done! Sorry for the hold up.

iangilman commented:
@AaronPlave Awesome, thank you for helping with this! 

@pearcetm Thank you for making this happen!
